paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt

```diff
@@ -169,13 +173,53 @@ public class PaparazziPlugin @Inject constructor(
       val isRecordRun = project.objects.property(Boolean::class.java)
       val isVerifyRun = project.objects.property(Boolean::class.java)
 
+      val xmlTestOutputDir = project.objects.directoryProperty()
+      val htmlTestReportOutputDir = project.objects.directoryProperty()
+
       project.gradle.taskGraph.whenReady { graph ->
         isRecordRun.set(recordTaskProvider.map { graph.hasTask(it) })
         isVerifyRun.set(verifyTaskProvider.map { graph.hasTask(it) })
       }
 
-      val testTaskProvider =
-        project.tasks.withType(Test::class.java).named { it == "test$testVariantSlug" }
+      val reportPaparazziProvider = project.tasks.register("reportPaparazzi$testVariantSlug") {
+        it.group = VERIFICATION_GROUP
+        it.description = "Generate report with snapshot diffs for failures"
+
+        it.mustRunAfter(testTaskProvider)
+
+        it.inputs.dir(buildDirectory.dir("paparazzi/failures"))
+          .withPropertyName("paparazzi.snapshot.failures")
+          .withPathSensitivity(PathSensitivity.RELATIVE)
+
+        it.inputs.dir(htmlTestReportOutputDir)
+          .withPropertyName("test.html.report.dir")
+          .withPathSensitivity(PathSensitivity.RELATIVE)
+
+        it.inputs.dir(xmlTestOutputDir)
+          .skipWhenEmpty()
+          .withPropertyName("test.result.xml.dir")
+          .withPathSensitivity(PathSensitivity.RELATIVE)
+
+        it.outputs.dir(snapshotReportOutputDir)
+          .withPropertyName("paparazzi.snapshot.report.output.dir")
+
+        it.doLast {
+          val failureSnapshotsDir = File(buildDirectory.asFile.get(), "paparazzi${separator}failures")
+
+          val reportDir = snapshotReportOutputDir.get().asFile
+          val resultsDir = xmlTestOutputDir.get().asFile
+
+          val report = app.cash.paparazzi.gradle.reporting.TestReport(
+            resultDir = resultsDir,
+            reportDir = reportDir,
+            failureSnapshotDir = failureSnapshotsDir,
+            applicationId = variant.namespace.get(),
+            variantKey = variant.name
+          )
+          report.generateScreenshotTestReport()
```

Comment:
move into separate test reporting task?

Reply:
its already in a separate task 🤔 ? `reportPaparazzi{Variant}UnitTest`

Reply:
sorry, i wasn't clear, and actually I was combining two questions:

1) Most important: Is there a way to customize the currently-existing Junit test reporting task?  A quick glance leads me to think that might be possible by configuring a custom [TestReport](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.TestReport.html) appropriately.  Otherwise, I am very reluctant to copying the entire test reporting task internals just to show the failure image diff.

Reply:
ya unfortunately `TestReport` task doesn't have the configuration we need. I tried this initially and opted for some string appending.

Yep totally down with creating an explicit task class

Reply:
> Otherwise, I am very reluctant to copying the entire test reporting task internals just to show the failure image diff.
> ya unfortunately TestReport task doesn't have the configuration we need.

What's your take on filing a feature request with Gradle to make this happen?  And is the value prop of copying (and then supporting) this extra task/configuration to expose one failure diff image worthwhile?

Reply:
> Is there a way to customize the currently-existing Junit test reporting task?

I looked into this some more, and it looks like there is a way!

Here's how I coming to this conclusion:

If I add the following to `sample/build.gradle`:

```
gradle.taskGraph.whenReady {
  gradle.taskGraph.allTasks.each { task ->
    println "task: " + task.name + ", class: " + task.class.name
  }
}

gradle.taskGraph.afterTask { task ->
  println "  inputs: "
  task.inputs.files.each { println "    " + it }

  println "  outputs: "
  task.outputs.files.each { println "    " + it }
}
```

and run:
```
$ gw sample:rPD --rerun
```

I get:
```
task: testDebugUnitTest, class: com.android.build.gradle.tasks.factory.AndroidUnitTest_Decorated

> Task :sample:testDebugUnitTest

ComposeA11yTest > compositeItems STANDARD_ERROR
    Sep 04, 2024 12:53:54 AM app.cash.paparazzi.internal.PaparazziLogger logAndroidFramework
    INFO: Settings [3]: Can't get key animator_duration_scale from content://settings/global

  ...

  inputs:
    /Users/jrod/Development/paparazzi/sample/build/tmp/kotlin-classes/debugUnitTest/app/cash/paparazzi/sample/WidgetTest.class
    /Users/jrod/Development/paparazzi/sample/build/tmp/kotlin-classes/debugUnitTest/app/cash/paparazzi/sample/ResourcesTest$Locale.class
	...
  outputs:
    ...
    /Users/jrod/Development/paparazzi/sample/build/reports/tests/testDebugUnitTest <==== the directory where HTML reports are written
    ...
	
```

From there, setting breakpoints in AGP's `AndroidUnitTest` and Gradle's `Test`, `AbstractTestTask` seem to show where reporting happens at the [end of test execution](https://github.com/gradle/gradle/blob/master/platforms/software/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java#L533), in particular, how a custom `TestReporter` can [be provided](https://github.com/gradle/gradle/blob/master/platforms/software/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java#L598) to generate a [custom report](https://github.com/gradle/gradle/blob/master/platforms/software/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java#L624).

While `AbstractTestTask#setTestReporter` is [package-private](https://github.com/gradle/gradle/blob/master/platforms/software/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java#L247), it's fairly reasonable to work around that:

```
# PaparazziPlugin: 177

      val testTaskProvider =
        project.tasks.withType(Test::class.java).named { it == "test$testVariantSlug" }
      testTaskProvider.configureEach { test ->
        test.setTestReporter { _, reportDir -> println(reportDir) }

    ....
}

private fun <T : AbstractTestTask> T.setTestReporter(testReporter: TestReporter) {
  AbstractTestTask::class.java
    .getDeclaredMethod("setTestReporter", TestReporter::class.java).apply {
      isAccessible = true
      invoke(this@setTestReporter, testReporter)
    }
}
```

Unless I'm missing something, this should require minimal change to the AGP-copied rendering classes, while removing (most of?) "reportPaparazzi$testVariantSlug", which was duplicating all that glue code anyway.

Reply:
> Otherwise, I am very reluctant to copying the entire test reporting task internals just to show the failure image diff.

To clarify this a bit, I'm fine with copying custom report rendering bits, but duplicating the test task is less ideal.

Reply:
nice find, i was digging through gradle source and the api docs and they didn't expose a public way to do this. I will try this route and defince a custom test reporter and see if that works.

I might still have to use the custom TestReport class potentially but we will see. 🤞🏽

Reply:
`To clarify this a bit, I'm fine with copying custom report rendering bits, but duplicating the test task is less ideal.` 

Makes sense, I was able to use the reflection method to hook into the test reporting. Once issue that could come up is that this could effect any consumers doing their own test manipulation. To prevent this, I plan on only overriding the report when `verifyPaparazziDebug` is being run on the task graph.